### PR TITLE
Toast rework with crossplatform support

### LINKED ISSUE #1000: Rewrite the toasts implementation for crossplatform
The current implementation of toasts in chatterino is currently not that good since it doesn't support very basic features and isn't crossplatform which chatterino strives to be.

### LINKED ISSUE #1435: Streamer profile picture in the live notification in chatterino is broken
**Describe the bug**
Since 2.1.5 whenever a streamer goes live it's just the standard "broken" profile picture, it works flawlessly in 2.1.4

**To reproduce**
Install 2.1.5, go to settings and then "notifications", activate "Enable Toasts" and add a channel in the tab "Channel going live"

**Screenshots**
2.1.4 

2.1.5 

**Chatterino version**
Chatterino 2.1.5 (commit c6e1a1de2)

**Operating system**
Windows 10 Pro Version 1909 OS Build 18363.476

**Additional information**
I also tried the latest nightly versions, it's also broken there.

### PR DESCRIPTION
**Pull request by apa420**
Toasts for crossplatform.
Fixes #1000
Fixes #1435

### CODE REVIEW

**Comment by apa420** [comment]
If anyone has time and wants to try and help testing this feature write here.

File: lib/snorenotify.pri [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# Requires cmake doxygen extra-cmake-modules
```
Comment (fourtf) [inline lib/snorenotify.pri]: Why does it require doxygen?

Comment (apa420) [reply lib/snorenotify.pri]: I haven't worked on this for a lot of weeks, it's leftover code that should be removed

File: BUILDING_ON_LINUX.md [hunk 2]
```diff
@@ -9,7 +9,7 @@ Note on Qt version compatibility: If you are installing Qt from a package manage
 
 ## Arch Linux
 install [chatterino2-git](https://aur.archlinux.org/packages/chatterino2-git/) from the aur or build manually as follows:
-1. `sudo pacman -S qt5-base qt5-multimedia qt5-svg gst-plugins-ugly gst-plugins-good boost rapidjson`
+1. `sudo pacman -S qt5-base qt5-multimedia qt5-svg gst-plugins-ugly gst-plugins-good boost rapidjson extra-cmake-modules`
```
Comment (apa420) [inline BUILDING_ON_LINUX.md]: Here as well

File: src/controllers/notifications/NotificationController.cpp [hunk 3]
```diff
@@ -217,4 +215,59 @@ void NotificationController::removeFakeChannel(const QString channelName)
...
+    popupWindow_->mouseRelease.connect(
+        [finishNotification, timer, callback, this](QMouseEvent *event) {
+            if (event->button() == Qt::LeftButton)
+            {
+                timer->stop();
+                finishNotification();
+                callback();
+            }
+        });
+    popupWindow_->update();
```
Comment (pphop) [inline src/controllers/notifications/NotificationController.cpp]: This call is not needed.

File: src/singletons/Toasts.cpp [hunk 4]
```diff
@@ -63,196 +49,115 @@ QString Toasts::findStringFromReaction(
...
+                QPixmap avatar;
+                avatar.loadFromData(data);
+
+                getApp()->notifications->addNotification(
+                    makeLayout(
+                        avatar,
+                        QString("<b>" + channelName + "</b> just went live!"),
+                        bottomText),
+                    std::chrono::milliseconds(
+                        (int)getSettings()->notificationDuration * 1000),
```
Comment (pphop) [inline src/singletons/Toasts.cpp]: Should cast after multiply so it actually uses fractional part.

File: src/singletons/Toasts.cpp [hunk 5]
```diff
@@ -63,196 +49,115 @@ QString Toasts::findStringFromReaction(
...
+            })
+        .onError(
+            [channelName, bottomText, callback, this](auto result) -> bool {
+                getApp()->notifications->addNotification(
+                    makeLayout(
+                        getResources().icon,
+                        QString("<b>" + channelName + "</b> just went live!"),
+                        bottomText),
+                    std::chrono::milliseconds(
+                        (int)getSettings()->notificationDuration * 1000),
```
Comment (pphop) [inline src/singletons/Toasts.cpp]: Same as comment above.

**Comment by leon-richardt** [comment]
Looks great so far :+1: Just wondering if it also makes sense to look at [QSystemTrayIcon](https://doc.qt.io/qt-5/qsystemtrayicon.html)?

File: src/widgets/dialogs/NotificationPopup.hpp [hunk 6]
```diff
@@ -2,28 +2,22 @@
...
-
-class Channel;
-using ChannelPtr = std::shared_ptr<Channel>;
+#include <QPixmap>
 
-struct Message;
-using MessagePtr = std::shared_ptr<const Message>;
+namespace chatterino {
 
 class NotificationPopup : public BaseWindow
```
Comment (leon-richardt) [inline src/widgets/dialogs/NotificationPopup.hpp]: Should this inherit from `BasePopup`?

**Comment by apa420** [comment]
> Looks great so far +1 Just wondering if it also makes sense to look at [QSystemTrayIcon](https://doc.qt.io/qt-5/qsystemtrayicon.html)?

Unsure how a system tray icon is connected to this feature. Please explain if you want.

I've refrained from using the system tray messages since they're more or less an ancient part of computer notification systems. Hence why there's a toast instead. I could see this system try feature being useful the same way as discord does it i.e. minimize to system tray instead of closing the application.

**Comment by leon-richardt** [comment]
> same way as discord does it

Yes, that's kind of the way I imagined it to work, e.g. showing a red dot in a system tray icon when there's a toast/unread notification pending. I agree that it can be argued to be more of a separate feature though.

**Comment by pajlada** [comment]
Looks good to me, would like you to bring your branch up to date with master first though.

File: src/controllers/notifications/NotificationController.cpp [hunk 7]
```diff
@@ -216,4 +215,58 @@ void NotificationController::removeFakeChannel(const QString channelName)
...
+            return;
+        }
+
+        startNotification();
+    };
+    popupWindow_->updatePosition();
+
+    if (auto layout = popupWindow_->layout(); layout)
+    {
+        qDeleteAll(popupWindow_->findChildren<QWidget *>(
```
[the remark this answers is not in the document]
Comment (apa420) [reply src/controllers/notifications/NotificationController.cpp]: The usage of `deleteLater` causes multiple toasts after each other to not display correctly.
Example: https://i.nuuls.com/Cv2Lj.mp4

**Comment by fourtf** [comment]
make a new one

**Pull request closed without merging**